File: rust/src/client/configuration/mobility_configuration.rs
/*
 * Software Name : libits-client
 * SPDX-FileCopyrightText: Copyright (c) Orange SA
 * SPDX-License-Identifier: MIT
 *
 * This software is distributed under the MIT license,
 * see the "LICENSE.txt" file for more details or https://opensource.org/license/MIT/
 *
 * Authors: see CONTRIBUTORS.md
 */

use ini::Properties;

use crate::client::configuration::configuration_error::ConfigurationError;
use crate::client::configuration::get_mandatory_from_properties;

pub(crate) const MOBILITY_SECTION: &str = "mobility";

const STATION_ID_FIELD: &str = "station_id";
const SOURCE_UUID_FIELD: &str = "source_uuid";
const USE_RESPONSIBILITY_FIELD: &str = "use_responsibility";

const THREAD_COUNT_FIELD: &str = "thread_count";

#[derive(Clone, Debug, Default)]
pub struct MobilityConfiguration {
    pub source_uuid: String,
    pub station_id: u32,
    pub use_responsibility: bool,
    pub thread_count: usize,
}

impl TryFrom<&Properties> for MobilityConfiguration {
    type Error = ConfigurationError;

    fn try_from(properties: &Properties) -> Result<Self, Self::Error> {
        let s = MobilityConfiguration {
            source_uuid: get_mandatory_from_properties(SOURCE_UUID_FIELD, properties)?,
            station_id: get_mandatory_from_properties(STATION_ID_FIELD, properties)?,
            use_responsibility: get_mandatory_from_properties(
                USE_RESPONSIBILITY_FIELD,
                properties,
            )?,
            thread_count: get_mandatory_from_properties(THREAD_COUNT_FIELD, properties)?,
        };
        Ok(s)
    }
}

File: swift/ITSClient/Sources/ITSMobility/MobilityConfiguration.swift
/*
 * Software Name : ITSClient
 * SPDX-FileCopyrightText: Copyright (c) Orange SA
 * SPDX-License-Identifier: MIT
 *
 * This software is distributed under the MIT license,
 * see the "LICENSE.txt" file for more details or https://opensource.org/license/MIT/
 *
 * Software description: Swift ITS client.
 */

import Foundation
import ITSCore

/// A structure to configure the mobility.
public struct MobilityConfiguration: Sendable {
    /// The `CoreConfiguration` which includes MQTT and telemetry configurations.
    public let coreConfiguration: CoreConfiguration
    /// The station identifier.
    public let stationID: UInt32
    /// The namespace used in the MQTT topic.
    public let namespace: String
    /// The user identifier.
    public var userIdentifier: String {
        coreConfiguration.mqttClientConfiguration.clientIdentifier
    }

    /// Initializes a `MobilityConfiguration`.
    /// - Parameters:
    ///   - coreConfiguration: The `CoreConfiguration` which includes MQTT and telemetry configurations.
    ///   - stationID: The station identifier. Must be unique and the same each time on a same device.
    ///   - namespace: The namespace used in the MQTT topic.
    public init(
        coreConfiguration: CoreConfiguration,
        stationID: UInt32,
        namespace: String = "default",
    ) {
        self.coreConfiguration = coreConfiguration
        self.stationID = stationID
        self.namespace = namespace
    }
}
